### PR TITLE
FIX: colormapping of integer normed images

### PR DESCRIPTION
**Pull request by tacaswell**
This is a work in progress and will hopefully be a starting point for @mdboom to fix it properly :smile: 

 - Use the same trick for carrying through the masked points to use the green and blue channels to carry the under/over markers.

Starts to address #6509

### CODE REVIEW

**Comment by mdboom** [comment]
Tried to make a PR against this, but we have too many contributors... ? ;)

Check out my branch for an additional fix for boundary norm: https://github.com/mdboom/matplotlib/tree/image-over-under

**Comment by tacaswell** [comment]
You have to just edit the URL.  They are aware of the problem, conclude it affects about 15 projects and oh well.

**Comment by efiring** [comment]
@tacaswell, https://github.com/efiring/matplotlib/tree/tacaswell-fix_over_under_images has my proposed fix on top of yours.  I don't quite understand the procedure for handling such combinations of commits; maybe it is simplest at this point for you to pull my commit from my tree and add it to this PR.  I haven't done any testing other than verifying that it makes 'image_masked.py' work.

**Comment by tacaswell** [comment]
@efiring @mdboom I have pulled Eric's commits into this branch (and re-based onto current v2.x).

File: lib/matplotlib/image.py [hunk 1]
```diff
@@ -356,21 +356,42 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
                 else:
-                    rgba[..., 3] = np.where(A.mask, 0, np.iinfo(A.dtype).max)
-                A = rgba
-                output = np.zeros((out_height, out_width, 4), dtype=A.dtype)
-                alpha = 1.0
-                created_rgba_mask = True
-            elif A.ndim == 3:
+                    # colormap norms that output integers (ex NoNorm
+                    # and BoundaryNorm) to RGBA space before
+                    # interpolating.  This is needed due to the the
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: Two 'the's.

File: lib/matplotlib/image.py [hunk 2]
```diff
@@ -356,21 +356,42 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
+                    # Agg resampler only working on floats in the
+                    # range [0, 1] and because interpolating indexes
+                    # into an arbitrary LUT maybe problematic.
+                    #
+                    # This falls back to interpolating in RGBA space which
+                    # can produce it's own artifacts of colors not in the map
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: it's -> its

File: lib/matplotlib/image.py [hunk 3]
```diff
@@ -356,21 +356,42 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
...
-                output = np.zeros((out_height, out_width, 4), dtype=A.dtype)
-                alpha = 1.0
-                created_rgba_mask = True
-            elif A.ndim == 3:
+                    # colormap norms that output integers (ex NoNorm
+                    # and BoundaryNorm) to RGBA space before
+                    # interpolating.  This is needed due to the the
+                    # Agg resampler only working on floats in the
+                    # range [0, 1] and because interpolating indexes
+                    # into an arbitrary LUT maybe problematic.
```
Comment (QuLogic) [inline lib/matplotlib/image.py]: maybe -> may be

**Comment by jenshnielsen** [comment]
Tests would be nice. At least an image comparison test based on the example in #6509 seems like a good idea

**Comment by tacaswell** [comment]
I'll add tests + fix the text tonight.

**Comment by mdboom** [comment]
I've said this "in person", but for the written record -- this seems like the right approach to me.  I'm :+1: on merging this, modulo everyone else's comments above.

**Pull request merged**